Analytical_Engine/Query/IsRectangle.cs

```diff
@@ -0,0 +1,113 @@
+using BH.oM.Geometry;
+using BH.oM.Reflection.Attributes;
+using BH.Engine.Geometry;
+using BH.Engine.Reflection;
+using BH.Engine.Spatial;
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.ComponentModel;
+
+namespace BH.Engine.Analytical
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Determines whether a panel is a rectangle")]
+        [Input("panel", "The Panel to check if it is a rectangle")]
+        [Output("bool", "True for rectangular panels or false for non-rectangular panels")]
+        public static bool IsRectangle<TEdge, TOpening>(this IPanel<TEdge, TOpening> panel)
+            where TEdge : IEdge
+            where TOpening : IOpening<TEdge>
+        {
+            //Check if there are any openings
+            if (panel.Openings.Count != 0)
+                return false;
+
+            List<ICurve> curves = panel.ExternalEdges.SelectMany(x => x.Curve.ISubParts()).ToList();
+
+            List<PolyCurve> polycurves = Engine.Geometry.Compute.IJoin(curves);
+
+            //Check there is a single continuous curve defining the Panel
+            if (polycurves.Count != 1)
+                return false;
+
+            PolyCurve polycurve = polycurves.First();
+
+            //Check that all subparts of the curve are linear
+            if (polycurve.SubParts().Any(x => !x.IIsLinear()))
+                return false;
+
+            //Group curves by direction vector to obtain discontinuity points
+            List<Point> points = new List<Point>();
+            var groupedCurves = curves.GroupBy(x => x.IStartDir());
+            foreach (var groupedCurve in groupedCurves)
+            {
+                ICurve jointCurve = Engine.Geometry.Compute.IJoin(groupedCurve.Select(x => x).ToList()).First();
+                points.Add(jointCurve.IStartPoint());
+            }
+
+            //Check there are four discontinuity points present (start and end same point)
+            if (points.Count != 4)
+                return false;
+
+            //Create vectors for all four sides of the quadilateral
+            List<Vector> vectors = new List<Vector>();
+            for (int i = 0; i < 3; i++)
+            {
+                vectors.Add(points[i + 1] - points[i]);
+            }
+            vectors.Add(points[0] - points[3]);
+
+            //Get the angles in the panel, only three are needed
+            List<double> angles = new List<double>() { vectors[3].Angle(vectors[0]) };
+            for (int i = 0; i < 3; i++)
+            {
+                angles.Add(vectors[i].Angle(vectors[i + 1]));
+            }
+
+            //Check the three angles are pi/2 degrees within tolerance
+            if (angles.Any(x => Math.Abs(Math.PI / 2 - x) > Tolerance.Angle))
```

Comment:
Actually think this test is enough. As long as the curves are planar (which they should be) and all angles are 90 degrees, I this is is a rectangle (or a square), and I would say that a square should return true for this as well, being a special case of a rectangle.

Reply:
Yeah agree on square being a special rectangle.